plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1914,12 +1748,223 @@ bool producer_plugin_impl::remove_expired_blacklisted_trxs( const fc::time_point
    return !exhausted;
 }
 
+// Returns contract name, action name, and exception text of an exception that occurred in a contract
+inline std::string get_detailed_contract_except_info(const packed_transaction_ptr& trx,
+                                                     const transaction_trace_ptr& trace,
+                                                     const fc::exception_ptr& except_ptr)
+{
+   std::string contract_name;
+   std::string act_name;
+   std::string details;
+
+   if( trace && !trace->action_traces.empty() ) {
+      auto last_action_ordinal = trace->action_traces.size() - 1;
+      contract_name = trace->action_traces[last_action_ordinal].receiver.to_string();
+      act_name = trace->action_traces[last_action_ordinal].act.name.to_string();
+   } else if ( trx ) {
+      const auto& actions = trx->get_transaction().actions;
+      if( actions.empty() ) return details; // should not be possible
+      contract_name = actions[0].account.to_string();
+      act_name = actions[0].name.to_string();
+   }
+
+   details = except_ptr ? except_ptr->top_message() : (trace && trace->except) ? trace->except->top_message() : std::string();
+   if (!details.empty()) {
+      details = fc::format_string("${d}", fc::mutable_variant_object() ("d", details), true);  // true for limiting the formatted string size
+   }
+
+   return "action: " + contract_name + ":" + act_name + ", " + details;
+}
+
+void producer_plugin_impl::log_trx_results( const transaction_metadata_ptr& trx,
+                                            const transaction_trace_ptr& trace, const fc::exception_ptr& except_ptr,
+                                            const fc::time_point& start )
```

Comment:
Create a version that only takes a `trace` and a version that only takes a `except_ptr` since they are never both provided.

Reply:
done